### ISSUE TITLE
[BUG] Incompatible with DnD5e V4

### ISSUE THREAD

**Issue by maxobremer** [MangoFVTT/fvtt-ready-set-roll-5e#489]
**Description**
The coming version of changed quit a bit and is not compatible with the current version of RSR. Some errors of note:

- Using an item/activity: The `dnd5e.preUseItem` hook has been deprecated and replaced with `dnd5e.preUseActivity'
- Also from using an item (spell, cantrip, weapon or sneak attack): Uncaught (in promise) TypeError: usageConfig.consume?.resources.includes is not a function
- Damage roll: The `dnd5e.preRollDamage` hook has been deprecated and replaced with `dnd5e.preRollDamageV2`
- Saving throw from a chat card: You are accessing the global "Die" which is now namespaced under foundry.dice.terms.Die
Deprecated since Version 12
- Using a loot item (which cant have activities, example being a bedroll): Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'size')

Rolling saves, skills and ability checks from the character sheet itself seems completely fine. Rolling anything from an item seems to not work
Also small note, initiative rolls seem to roll as core foundry rolls, their chat cards don't use the stylising of RSR.

**Steps to Reproduce**
1. Install the prerelease version [4.0.0](https://github.com/foundryvtt/dnd5e/releases/tag/release-4.0.0) of dnd5e 
2. Enable RSR
3. Try to roll some attacks with a starter hero

**Versions**
 - Module: [3.2.0]
 - dnd5e: [4.0.0]
 - Foundry: [12.331 (node version)]

**Other Modules**
None

**Comment by crash1115** [comment]
One thing that might be worth considering during this update is allowing users to select which individual types of items or activities quick roll. There were a couple changes and improvements to certain roll dialogs that users might want to have access to, while still fast forwarding other roll types. Off the top of my head:
- A selector for 1-handed/2-handed for versatile, thrown, and offhand attacks appears in the attack dialog
- The damage dialog has a situational bonus field for each type of damage now

Sample use case:
Someone rocking a versatile weapon might want attacks to roll normally so they can get the 1h/2h selector in the attack dialog, but have damage, spells and utility activities fast forward.

**Comment by oestred** [comment]
@MangoFVTT I love your module, definitely a must have for playing.
Is there any update you can share yet about RSR getting updated for D&D 4.x?

**Comment by MangoFVTT** [comment]
> Is there any update you can share yet about RSR getting updated for D&D 4.x?

I'm still alive, but my past month has been extremely busy with work for my job and house so I haven't had any real time to do extra development for RSR. As it stands, I haven't started anything for 4.x yet, but I hope to get to it when things calm down in my personal life. No real timeline for that, I hope that it's soon 😅

**Comment by gludington** [comment]
The Uncaught (in promise) TypeError: usageConfig.consume?.resources.includes is not a function

I noticed when looking at this one: https://github.com/foundryvtt/dnd5e/issues/4229

So it may be an issue in the system level hook that needs addressing

**Comment by asherett** [comment]
> Using a loot item (which cant have activities, example being a bedroll): Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'size')
This was a system bug, fixed now.

**Comment by friction1** [comment]
Not sure if this helps, but my damage roll isn't appearing either and this is the error that shows up in the log. 
<img width="401" alt="Screenshot 2024-10-05 at 10 46 40 AM" src="https://github.com/user-attachments/assets/89190304-206a-4a35-9bce-a3c969641011">

**Comment by gludington** [comment]
Discussion on rolling workflow in the 5e system:

https://github.com/foundryvtt/dnd5e/issues/4459

**Comment by jagoe** [comment]
@MangoFVTT Is there anything one could do to help? I'm not familiar with your code base, but I'm familiar with Foundry module development in general. If there's specific sections in the code that you could point me to, I might be able to provide some support.

**Comment by MangoFVTT** [comment]
@jagoe normally, picking up any issue in the issues list and making a PR for it is always appreciated. At the very least it would do the groundwork for me. However, since this is a major release, there are some majorly sweeping changes that need to happen which make it difficult for anyone else to contribute in a way that speeds things up for me. I do very much appreciate the offer though ❤️

**Comment by jagoe** [comment]
Yeah, I thought as much :sweat_smile: Feel free to hit me up should you find anything or for testing. In the meantime, I'll take a look at the issues list. Though I expect it would make more sense to grab something from there once you're done with the overhaul.

**Comment by MangoFVTT** [comment]
I have pushed v3.3.0 that should bring compatibility with dnd5e 4.0+, which means this ticket is resolved. Since this was a major overhaul in a time crunch, I do expect plenty of bugs. These should be reported as separate issues moving forward

**Comment by oestred** [comment]
I just wanted to say thank you and that all your effort is appreciated!  I know RSR is very popular because it is such an awesome module.  
I'll check it out on my test install and if it looks good use it this week.  This week is our first week starting our new 2024 campaign so it works out perfect. I'll open an issue if any are run into.
